storage/Devices/DeviceImpl.h

```diff
@@ -21,6 +21,32 @@ namespace storage
     class SystemInfo;
 
 
+    template <typename Type>
+    Type* to_device_of_type(Device* device)
+    {
+	static_assert(!is_const<Type>::value, "Type must not be const");
+
+	Type* tmp = dynamic_cast<Type*>(device);
+	if (!tmp)
+	    ST_THROW(DeviceHasWrongType("device has wrong type"));
```

Comment:
Unfortunatelty that gives an error message like 'expected a N7storage10FilesystemE'.

So traits will be needed to get nice names.

Apart from that we can even go further:

```ST_THROW(DeviceHasWrongType(sformat("device has wrong type, expected '%s' but seen '%s'", typeid(Type).name(), device->get_impl().get_classname())));```